Blazor.ExtraDry/Blazor.ExtraDry.Core/Rules/RuleEngine.cs

```diff
@@ -138,36 +138,91 @@ private async Task<object> ResolveEntityValue(PropertyInfo property, object sour
         [SuppressMessage("Performance", "CA1822:Mark members as static", Justification = "Keep as standard service instance style for DI.")]
-        public void Delete<T>(T item, Action hardDeleteOption = null)
+        public void DeleteSoft<T>(T item, Action hardDeletePrepare, Action hardDeleteCommit)
         {
             if(item == null) {
                 throw new ArgumentNullException(nameof(item));
             }
-            var properties = typeof(T).GetProperties();
+            if(!AttemptSoftDelete(item)) {
+                if(hardDeletePrepare == null) {
+                    throw new InvalidOperationException("Item could not be soft-deleted and no hard delete action was provided.");
+                }
+                else {
+                    hardDeletePrepare();
+                    hardDeleteCommit?.Invoke();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Processes a hard delete of an item if possible, even if a `DeleteValue` is provided in the `RulesAttribute`.
+        /// This allows for the database to notify if a hard-delete is not possible by throwing an exception in the commit.
+        /// If not possible, then the soft-delete is performed instead.
+        /// If neither is possible, an exception is thrown.
+        /// </summary>
+        /// <param name="item">The item to delete, a soft-delete is attempted first.</param>
+        /// <param name="hardDeletePrepare">If item can't be soft-deleted, then the action that is executed for a hard-delete.</param>
+        /// <param name="hardDeleteCommit">If item can't be soft-deleted, then the optional action that is executed to commit hard-delete.</param>
+        [SuppressMessage("Performance", "CA1822:Mark members as static", Justification = "Keep as standard service instance style for DI.")]
+        public void DeleteHard<T>(T item, Action hardDeletePrepare, Action hardDeleteCommit)
+        {
+            if(item == null) {
+                throw new ArgumentNullException(nameof(item));
+            }
+            if(hardDeletePrepare == null) {
+                throw new ArgumentNullException(nameof(item), "Must provide an action to prepare hard delete, e.g. remove from collection");
+            }
+            if(hardDeleteCommit == null) {
+                throw new ArgumentNullException(nameof(item), "Must provide an action to commit hard delete, e.g. save to database");
+            }
+            try {
+                hardDeletePrepare();
+                hardDeleteCommit();
+            }
+            catch { // Hmmm, can we be specific about exception type here?
+                if(!AttemptSoftDelete(item)) {
+                    throw new InvalidOperationException("Item could not be hard-deleted and no rules for soft-delete provided.");
+                }
+                hardDeleteCommit();
```

Comment:
Should this be soft delete? It's trying to hard delete, then if it fails, it tries again?

Reply:
Perhaps the naming is confusing the issue.  The idea of the function is that it does and commits a hard-delete.  If that fails it does and commits the soft-delete.  I wanted to ensure that the end of this function left you with the data in the database without the need to check a result and decide if you still needed to .SaveChangesAsync().

Reply:
Yeppo, it'll be hte naming that really confused me. That's fine then, maybe a comment if you're still in this space later